tools/advisor/advisor/db_config_optimizer.py

```diff
@@ -0,0 +1,385 @@
+            action = suggestions_dict[sugg_name].action
+            if not (option and action):  # suggestion not in required format
+                continue
+            required_options.append(option)
+            rule_suggestions.append(suggestions_dict[sugg_name])
+        current_config = options.get_options(required_options)
+        # Create the updated configuration from the rule's suggestions
+        updated_config = {}
+        for sugg in rule_suggestions:
+            # case: when the option is not present in the current configuration
+            if sugg.option not in current_config:
+                new_value = ConfigOptimizer.apply_action_on_value(
+                    None, sugg.action, sugg.suggested_values
+                )
+                if new_value:
+                    if sugg.option not in updated_config:
+                        updated_config[sugg.option] = {}
+                    if '.' not in sugg.option:
+                        updated_config[sugg.option][NO_FAM] = new_value
+                    else:
+                        for col_fam in rule.get_trigger_column_families():
+                            updated_config[sugg.option][col_fam] = new_value
+                continue
+            # case: when the option is present in the current configuration
+            if NO_FAM in current_config[sugg.option]:
+                old_value = current_config[sugg.option][NO_FAM]
+                new_value = ConfigOptimizer.apply_action_on_value(
+                    old_value, sugg.action, sugg.suggested_values
+                )
+                if new_value:
+                    if sugg.option not in updated_config:
+                        updated_config[sugg.option] = {}
+                    updated_config[sugg.option][NO_FAM] = new_value
+            else:
+                for col_fam in rule.get_trigger_column_families():
+                    old_value = None
+                    if col_fam in current_config[sugg.option]:
+                        old_value = current_config[sugg.option][col_fam]
+                    new_value = ConfigOptimizer.apply_action_on_value(
+                        old_value, sugg.action, sugg.suggested_values
+                    )
+                    if new_value:
+                        if sugg.option not in updated_config:
+                            updated_config[sugg.option] = {}
+                        updated_config[sugg.option][col_fam] = new_value
+        return current_config, updated_config
+
+    @staticmethod
+    def pick_rule_to_apply(rules, current_rule_name, rules_tried, backtrack):
+        if not rules:
+            print('\nNo more rules triggered!')
+            return None
+        if current_rule_name and not backtrack:
+            for rule in rules:
+                if rule.name == current_rule_name:
+                    return rule
+        for rule in rules:
+            if rule.name not in rules_tried:
```

Comment:
Well for increase suggestions we increase the value by 1.3x. So it is reasonable to apply the rule again with another 1.3x increase. Can you put a TODO to clarify that the current policy is just some initial and should be changed in future iterations.

Reply:
We pick the same rule again as long as it gives us improvement in the throughput.
However, if backtrack is TRUE, which means that the throughput decreased after applying that rule, we try to pick another not-yet-tried rule.

Reply:
can you explain that with inline comments?